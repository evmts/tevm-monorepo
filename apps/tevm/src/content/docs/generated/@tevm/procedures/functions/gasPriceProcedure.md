---
editUrl: false
next: false
prev: false
title: "gasPriceProcedure"
---

> **gasPriceProcedure**(`options`): `EthGasPriceJsonRpcProcedure`

## Parameters

▪ **options**: `object`

▪ **options.blockchain**: `Blockchain`

▪ **options.forkUrl**: `undefined` \| `string`

## Returns

## Source

[vm/procedures/src/jsonrpc/eth/gasPriceProcedure.js:9](https://github.com/evmts/tevm-monorepo/blob/main/vm/procedures/src/jsonrpc/eth/gasPriceProcedure.js#L9)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
