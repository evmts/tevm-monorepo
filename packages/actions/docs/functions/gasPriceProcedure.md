[**@tevm/actions**](../README.md) • **Docs**

***

[@tevm/actions](../globals.md) / gasPriceProcedure

# Function: gasPriceProcedure()

> **gasPriceProcedure**(`options`): [`EthGasPriceJsonRpcProcedure`](../type-aliases/EthGasPriceJsonRpcProcedure.md)

## Parameters

• **options**: `TevmNode`\<`"fork"` \| `"normal"`, `object`\>

## Returns

[`EthGasPriceJsonRpcProcedure`](../type-aliases/EthGasPriceJsonRpcProcedure.md)

## Defined in

[packages/actions/src/eth/gasPriceProcedure.js:9](https://github.com/evmts/tevm-monorepo/blob/main/packages/actions/src/eth/gasPriceProcedure.js#L9)
