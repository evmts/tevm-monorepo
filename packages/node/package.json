{
	"name": "@tevm/node",
	"version": "2.0.0-next.108",
	"private": false,
	"description": "Client implementations of Tevm",
	"keywords": [
		"solidity",
		"sol",
		"typescript",
		"web3",
		"blockchain"
	],
	"repository": {
		"type": "git",
		"url": "https://github.com/evmts/tevm-monorepo.git",
		"directory": "core/node"
	},
	"license": "MIT",
	"contributors": [
		"Will Cory <willcory10@gmail.com>"
	],
	"type": "module",
	"exports": {
		"./package.json": "./package.json",
		".": {
			"import": {
				"types": "./dist/index.d.ts",
				"default": "./dist/index.js"
			},
			"require": {
				"types": "./dist/index.d.cts",
				"default": "./dist/index.cjs"
			}
		}
	},
	"main": "dist/index.cjs",
	"module": "dist/index.js",
	"types": "dist/index.d.ts",
	"files": [
		"dist",
		"src",
		"!src/**/*.spec.ts",
		"types"
	],
	"scripts": {
		"all": "pnpm i  && bun run build && bun lint && bun format && bun test:run && bun generate:docs && bun lint:deps && bun lint:package",
		"build": "nx run-many --targets=build:dist,build:types --projects=@tevm/node",
		"build:dist": "tsup",
		"build:types": "tsup --dts-only",
		"clean": "rm -rf node_modules && rm -rf artifacts && rm -rf dist && rm -rf cache",
		"format": "biome format . --write",
		"format:check": "biome format .",
		"generate:docs": "typedoc",
		"lint": "biome check . --write --unsafe",
		"lint:check": "biome check . --verbose",
		"lint:deps": "bunx depcheck",
		"lint:package": "bunx publint --strict && attw --pack",
		"package:up": "pnpm up --latest",
		"test": "vitest --watch",
		"test:coverage": "vitest --coverage",
		"test:run": "vitest",
		"typecheck": "tsc --noEmit"
	},
	"dependencies": {
		"@tevm/address": "workspace:^",
		"@tevm/block": "workspace:^",
		"@tevm/blockchain": "workspace:^",
		"@tevm/common": "workspace:^",
		"@tevm/evm": "workspace:^",
		"@tevm/jsonrpc": "workspace:^",
		"@tevm/logger": "workspace:^",
		"@tevm/predeploys": "workspace:^",
		"@tevm/receipt-manager": "workspace:^",
		"@tevm/state": "workspace:^",
		"@tevm/sync-storage-persister": "workspace:^",
		"@tevm/tx": "workspace:^",
		"@tevm/txpool": "workspace:^",
		"@tevm/utils": "workspace:^",
		"@tevm/vm": "workspace:^"
	},
	"devDependencies": {
		"@tevm/test-utils": "workspace:^",
		"@tevm/tsconfig": "workspace:^",
		"@tevm/tsupconfig": "workspace:^"
	},
	"peerDependencies": {
		"viem": "^2.14.2"
	},
	"publishConfig": {
		"access": "public"
	},
	"sideEffect": false
}
