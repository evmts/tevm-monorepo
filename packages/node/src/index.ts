export { createTevmNode } from './createTevmNode.js'
export type { TevmNode } from './TevmNode.js'
export type { TevmNodeOptions } from './TevmNodeOptions.js'
export type { BaseClientOptions } from './BaseClientOptions.js'
export type { CustomPrecompile } from './CustomPrecompile.js'
export type { Extension } from './Extension.js'
export type { Hardfork } from './Hardfork.js'
export type {
	MiningConfig,
	AutoMining,
	ManualMining,
	IntervalMining,
} from './MiningConfig.js'
export { prefundedAccounts, GENESIS_STATE } from './GENESIS_STATE.js'
export type { Filter, FilterType } from './Filter.js'
export {
	ProviderRpcError,
	type EIP1193Events,
	type EIP1193EventMap,
	type ProviderMessage,
	type EIP1193EventEmitter,
	type ProviderConnectInfo,
} from './EIP1193EventEmitterTypes.js'
