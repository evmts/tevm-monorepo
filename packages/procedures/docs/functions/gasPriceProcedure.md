**@tevm/procedures** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > gasPriceProcedure

# Function: gasPriceProcedure()

> **gasPriceProcedure**(`options`): `EthGasPriceJsonRpcProcedure`

## Parameters

▪ **options**: `Pick`\<`object`, `"getVm"` \| `"forkUrl"`\>

## Returns

## Source

[procedures/src/eth/gasPriceProcedure.js:9](https://github.com/evmts/tevm-monorepo/blob/main/packages/procedures/src/eth/gasPriceProcedure.js#L9)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
