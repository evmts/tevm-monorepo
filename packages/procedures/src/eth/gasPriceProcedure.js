import { gasPriceHandler } from '@tevm/actions'
import { numberToHex } from '@tevm/utils'

/**
 * @param {Parameters<typeof gasPriceHandler>[0]} options
 * @returns {import('@tevm/procedures-types').EthGasPriceJsonRpcProcedure}
 */
export const gasPriceProcedure =
	({ vm, forkUrl }) =>
	async (req) => ({
		...(req.id ? { id: req.id } : {}),
		jsonrpc: '2.0',
		method: req.method,
		result: await gasPriceHandler({ vm, forkUrl })({}).then(numberToHex),
	})
