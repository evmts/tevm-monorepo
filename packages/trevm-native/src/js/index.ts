import { createLogger } from '@tevm/logger';
import { InternalError } from '@tevm/errors';
import type { StateManager } from '@tevm/state';
import type { Common } from '@tevm/common';
import type { Chain } from '@tevm/blockchain';

// Import native module
import { init_trevm, TrevmWrapper } from '../../dist/index.node';

export interface TrevmEvmOptions {
  common: Common;
  stateManager: StateManager;
  blockchain: Chain;
  customPrecompiles?: any[];
  profiler?: boolean;
  allowUnlimitedContractSize?: boolean;
  loggingLevel?: string;
}

export interface TrevmRunCallOptions {
  caller?: string;
  to?: string;
  data?: string;
  value?: bigint;
  gasLimit?: bigint;
  skipBalance?: boolean;
}

export interface TrevmRunCallResult {
  returnValue: string;
  executionGasUsed: string;
  error?: string;
}

export interface TrevmEvm {
  runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult>;
}

/**
 * Creates a trevm-based EVM instance
 * @param options EVM options
 * @returns A trevm EVM instance
 */
export const createTrevm = async ({
  common,
  stateManager,
  blockchain,
  customPrecompiles,
  profiler,
  allowUnlimitedContractSize,
  loggingLevel,
}: TrevmEvmOptions): Promise<TrevmEvm> => {
  const logger = createLogger({
    name: '@tevm/trevm',
    level: loggingLevel ?? 'warn',
  });
  
  logger.debug({
    allowUnlimitedContractSize,
    profiler,
    customPrecompiles: customPrecompiles?.map(c => c.address.toString()),
  });

  try {
    // Initialize the native trevm module with callbacks to the StateManager
    const nativeTrevm = await init_trevm({
      chainId: common.chainId(),
      allowUnlimitedContractSize: allowUnlimitedContractSize ?? false,
      stateCallbacks: {
        getAccount: async (address: string) => {
          try {
            const account = await stateManager.getAccount(address);
            if (!account) return null;
            
            return {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? `0x${Buffer.from(account.codeHash).toString('hex')}` : undefined,
              code: account.code ? `0x${Buffer.from(account.code).toString('hex')}` : undefined,
            };
          } catch (error) {
            logger.error(`Error in getAccount for ${address}:`, error);
            return null;
          }
        },
        putAccount: async (address: string, account: any) => {
          try {
            await stateManager.putAccount(address, {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? Buffer.from(account.codeHash.slice(2), 'hex') : undefined,
              code: account.code ? Buffer.from(account.code.slice(2), 'hex') : undefined,
            });
            return true;
          } catch (error) {
            logger.error(`Error in putAccount for ${address}:`, error);
            return false;
          }
        },
        getStorage: async (address: string, key: string) => {
          try {
            const value = await stateManager.getContractStorage(address, key);
            return value ? `0x${Buffer.from(value).toString('hex')}` : '0x0';
          } catch (error) {
            logger.error(`Error in getStorage for ${address} key ${key}:`, error);
            return '0x0';
          }
        },
        putStorage: async (address: string, key: string, value: string) => {
          try {
            await stateManager.putContractStorage(
              address,
              key,
              Buffer.from(value.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putStorage for ${address} key ${key}:`, error);
            return false;
          }
        },
        getCode: async (address: string) => {
          try {
            const code = await stateManager.getContractCode(address);
            return code ? `0x${Buffer.from(code).toString('hex')}` : '0x';
          } catch (error) {
            logger.error(`Error in getCode for ${address}:`, error);
            return '0x';
          }
        },
        putCode: async (address: string, code: string) => {
          try {
            await stateManager.putContractCode(
              address,
              Buffer.from(code.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putCode for ${address}:`, error);
            return false;
          }
        },
        getBlockHash: async (blockNumber: number) => {
          try {
            const block = await blockchain.getBlock(BigInt(blockNumber));
            return block ? `0x${Buffer.from(block.header.hash()).toString('hex')}` : null;
          } catch (error) {
            logger.error(`Error in getBlockHash for block ${blockNumber}:`, error);
            return null;
          }
        }
      }
    });

    // Return a JavaScript wrapper around the native trevm that matches the Evm interface
    return {
      async runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult> {
        try {
          return await nativeTrevm.run_call({
            caller: opts.caller,
            to: opts.to,
            data: opts.data,
            value: opts.value,
            gas_limit: opts.gasLimit,
            skip_balance: opts.skipBalance,
          });
        } catch (error) {
          logger.error('Error in runCall:', error);
          throw new InternalError(`Trevm execution error: ${error}`);
        }
      }
    };
  } catch (error) {
    logger.error('Error initializing trevm:', error);
    throw new InternalError(`Failed to initialize trevm: ${error}`);
  }
};