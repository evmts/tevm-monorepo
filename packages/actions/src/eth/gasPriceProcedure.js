import { numberToHex } from '@tevm/utils'
import { gasPriceHandler } from './gasPriceHandler.js'

/**
 * @param {Parameters<typeof gasPriceHandler>[0]} options
 * @returns {import('./EthProcedure.js').EthGasPriceJsonRpcProcedure}
 */
export const gasPriceProcedure =
	({ getVm, forkTransport }) =>
	async (req) => {
		try {
			// TODO pass in a client instead
			const result = await gasPriceHandler(/** @type any*/ ({ getVm, forkTransport }))({}).then(numberToHex)
			return {
				...(req.id ? { id: req.id } : {}),
				jsonrpc: '2.0',
				method: req.method,
				result,
			}
		} catch (error) {
			return {
				...(req.id ? { id: req.id } : {}),
				jsonrpc: '2.0',
				method: req.method,
				error: {
					code: -32000,
					message: error instanceof Error ? error.message : String(error),
				},
			}
		}
	}
