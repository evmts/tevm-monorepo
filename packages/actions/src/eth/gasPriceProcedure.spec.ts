import { type TevmNode, createTevmNode } from '@tevm/node'
import { beforeEach, describe, expect, it } from 'vitest'
import type { EthGasPriceJsonRpcRequest } from './EthJsonRpcRequest.js'
import { gasPriceProcedure } from './gasPriceProcedure.js'

let client: TevmNode

beforeEach(() => {
	client = createTevmNode()
})

describe('gasPriceProcedure', () => {
	it('should return the current gas price', async () => {
		const request: EthGasPriceJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_gasPrice',
			id: 1,
			params: [],
		}

		const response = await gasPriceProcedure({
			getVm: client.getVm,
			forkTransport: client.forkTransport,
		} as any)(request)

		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_gasPrice')
		expect(response.id).toBe(request.id as any)
		expect(response.result).toMatchSnapshot()
	})

	it('should handle requests without an id', async () => {
		const request: EthGasPriceJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_gasPrice',
			params: [],
		}

		const response = await gasPriceProcedure({
			getVm: client.getVm,
			forkTransport: client.forkTransport,
		} as any)(request)

		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_gasPrice')
		expect(response.id).toBeUndefined()
		expect(response.result).toMatchSnapshot()
	})
})
