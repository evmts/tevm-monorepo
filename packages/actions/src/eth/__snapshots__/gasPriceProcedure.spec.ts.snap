// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`gasPriceProcedure > should handle requests without an id 1`] = `"0x3b9aca00"`;

exports[`gasPriceProcedure > should return the current gas price 1`] = `"0x3b9aca00"`;
